sway-core/src/type_system/engine.rs

```diff
@@ -40,27 +154,649 @@ pub(crate) struct Unification {
+        let tsi = TypeSourceInfo {
+            type_info: unknown_generic.into(),
+            source_id,
+        };
+        TypeId::new(self.slab.insert(tsi))
+    }
+
+    /// Inserts a new [TypeInfo::UnknownGeneric] into the [TypeEngine]
+    /// that represents a `Self` type and returns its [TypeId].
+    /// The unknown generic `name` [Ident] will be set to "Self" with the provided `use_site_span`.
+    ///
+    /// Note that the span in general does not point to a reserved word "Self" in
+    /// the source code, nor is related to it. The `Self` type represents the type
+    /// in `impl`s and does not necessarily relate to the "Self" keyword in code.
+    ///
+    /// Therefore, *the span must always point to a location in the source file in which
+    /// the particular `Self` type is, e.g., being declared or implemented*.
+    ///
+    /// `is_from_type_parameter` denotes if the `Self` type is from type parameter or not.
+    pub(crate) fn new_unknown_generic_self(
+        &self,
+        use_site_span: Span,
+        is_from_type_parameter: bool,
+    ) -> TypeId {
+        let name = Ident::new_with_override("Self".into(), use_site_span);
+        self.new_unknown_generic(name.clone(), VecSet(vec![]), None, is_from_type_parameter)
+    }
+
+    /// Same as [Self::new_unknown_generic_self], except that, together with the [TypeId], it also returns the [Ident]
+    /// with the value "Self" and the provided `use_site_span`.
+    pub(crate) fn new_unknown_generic_self_with_name(
+        &self,
+        use_site_span: Span,
+        is_from_type_parameter: bool,
+    ) -> (TypeId, Ident) {
+        let name = Ident::new_with_override("Self".into(), use_site_span);
+        let type_id =
+            self.new_unknown_generic(name.clone(), VecSet(vec![]), None, is_from_type_parameter);
+        (type_id, name)
+    }
+
+    /// Inserts a new [TypeInfo::Enum] into the [TypeEngine] and returns
+    /// its [TypeId], or returns a [TypeId] of an existing shareable enum type
+    /// that corresponds to the enum given by the `decl_id`.
+    pub(crate) fn insert_enum(&self, engines: &Engines, decl_id: DeclId<TyEnumDecl>) -> TypeId {
+        let decl = engines.de().get_enum(&decl_id);
+        let source_id = Self::get_enum_fallback_source_id(&decl);
+        let is_shareable_type = self.is_shareable_enum(engines.de(), &decl);
+        let type_info = TypeInfo::Enum(decl_id);
+        self.insert_or_replace_type_source_info(
+            engines,
+            type_info,
+            source_id,
+            is_shareable_type,
+            None,
+        )
+    }
+
+    /// Inserts a new [TypeInfo::Struct] into the [TypeEngine] and returns
+    /// its [TypeId], or returns a [TypeId] of an existing shareable struct type
+    /// that corresponds to the struct given by the `decl_id`.
+    pub(crate) fn insert_struct(&self, engines: &Engines, decl_id: DeclId<TyStructDecl>) -> TypeId {
```

Comment:
Well, another one proof that devs do not read method and struct docs :smile:

Jokes aside, I thought of the same. Having API speaking for itself rather then explaining the usage intension in the docs.

Actually, I was thinking exactly about the same name, `get_or_insert` :wink: 

But I've opted in the end for leaving `insert` as the name, because of the following reasons:
- Starting from the least important reason. `get_or_insert_<type>[<options>]` would be much more verbose then only `insert_...`. The already existing `insert` method which keeps the original semantics, would also need to be changed to `get_or_insert`, and to the current name we are already well used.

- Having `get_or_insert...` would actually make the API more confusing. What are we inserting and what are we getting from the `TypeEngine`? It would make `TypeEngine` inconsistent with, e.g., `DeclEngine` in that regard. Let me explain. We are inserting types (`TypeInfo`) and have `get` method for retreiving, again `TypeInfo`s. Having suddenly an abundance of `get...` methods returning `TypeId`s makes it unclear what the `TypeEngine` is actually supposed to return when asked for `get`. This is for me the major reason for not having `get_or_insert`. The `TypeEngine` is for _getting_ `TypeInfo`s and not `TypeId`s. The `TypeId`s are just an acknowledgments that an _insert_ worked (more on this in the third point below).

- Finally, _inserting_ is what we are actually doing. I like to compare it here with the `HashSet`. The `insert` might create a new entry in the set or might not, and informs about that via the resulting `bool`. Similarly, our `insert` can return an existing `TypeId`, or a new one, but in both cases, we are attempting an _insert_. `HashSet::get_or_insert` properly returns `T` and not `bool`. This is what I also meant in the point above. If we `get...` we always get what the abstraction contains, `TypeInfo` in case of `TypeEngine`, or `T` in the case of the `HashSet`. `HashSet` respects this. Having `get_or_insert...` in the `TypeEngine` not returning `TypeInfo` but something else would break this.

These are the reasons why I've opted for keeping the name `insert...` for general inserting, and `new` and `id_of` for _surely adding_ and _surely retreiving a fixed id_. Accordingly, I've conveyed the intended usage in the code docs.

Does this approach makes sense to you?